lib/src/front_end/ast_node_visitor.dart

```diff
@@ -683,121 +683,36 @@ class AstNodeVisitor extends ThrowingAstVisitor<Piece> with PieceFactory {
-                initialization: null,
-                leftSeparator: Token(precedingComments: null),
-                condition: null,
-                rightSeparator: Token(precedingComments: null),
-                updaters: NodeList(isEmpty: true),
-              ) &&
-              var forParts
-          when node.rightParenthesis.precedingComments == null:
-        forPartsPiece = buildPiece((b) {
-          b.token(node.leftParenthesis);
-          b.token(forParts.leftSeparator);
-          b.token(forParts.rightSeparator);
-          b.token(node.rightParenthesis);
-        });
-
-      case ForParts forParts &&
-            ForPartsWithDeclarations(variables: AstNode? initializer):
-      case ForParts forParts &&
-            ForPartsWithExpression(initialization: AstNode? initializer):
-        // In a C-style for loop, treat the for loop parts like an argument list
-        // where each clause is a separate argument. This means that when they
-        // split, they split like:
-        //
-        // ```
-        // for (
-        //   initializerClause;
-        //   conditionClause;
-        //   incrementClause
-        // ) {
-        //   body;
-        // }
-        // ```
-        var partsList =
-            DelimitedListBuilder(this, const ListStyle(commas: Commas.none));
-        partsList.leftBracket(node.leftParenthesis);
-
-        // The initializer clause.
-        if (initializer != null) {
-          partsList.addCommentsBefore(initializer.beginToken);
-          partsList.add(buildPiece((b) {
-            b.visit(initializer);
-            b.token(forParts.leftSeparator);
-          }));
-        } else {
-          // No initializer, so look at the comments before `;`.
-          partsList.addCommentsBefore(forParts.leftSeparator);
-          partsList.add(tokenPiece(forParts.leftSeparator));
-        }
+    var forPartsPiece = createForLoopParts(
+        node.leftParenthesis, node.forLoopParts, node.rightParenthesis);
+    var body = nodePiece(node.body);
 
-        // The condition clause.
-        if (forParts.condition case var conditionExpression?) {
-          partsList.addCommentsBefore(conditionExpression.beginToken);
-          partsList.add(buildPiece((b) {
-            b.visit(conditionExpression);
-            b.token(forParts.rightSeparator);
-          }));
-        } else {
-          partsList.addCommentsBefore(forParts.rightSeparator);
-          partsList.add(tokenPiece(forParts.rightSeparator));
-        }
+    var forPiece = ForPiece(forKeyword, forPartsPiece, body,
+        hasBlockBody: node.body.isSpreadCollection);
 
-        // The update clauses.
-        if (forParts.updaters.isNotEmpty) {
-          partsList.addCommentsBefore(forParts.updaters.first.beginToken);
-          partsList.add(createList(forParts.updaters,
-              style: const ListStyle(commas: Commas.nonTrailing)));
-        }
+    // If the body of a for element is itself a control flow element, then
```

Comment:
[nit] This repeats the code - are the any "why" details worth adding?

Reply:
Good point done.